core/arch/arm/plat-imx/pm/pm-imx7.c

```diff
@@ -0,0 +1,264 @@
+	{ 0x0, READ_DATA_FROM_HARDWARE },
+	{ 0x4, READ_DATA_FROM_HARDWARE },
+	{ 0x10, READ_DATA_FROM_HARDWARE },
+	{ 0xb0, READ_DATA_FROM_HARDWARE },
+	{ 0x9c, READ_DATA_FROM_HARDWARE },
+	{ 0x7c, READ_DATA_FROM_HARDWARE },
+	{ 0x80, READ_DATA_FROM_HARDWARE },
+	{ 0x84, READ_DATA_FROM_HARDWARE },
+	{ 0x88, READ_DATA_FROM_HARDWARE },
+	{ 0x6c, READ_DATA_FROM_HARDWARE },
+	{ 0x20, READ_DATA_FROM_HARDWARE },
+	{ 0x30, READ_DATA_FROM_HARDWARE },
+	{ 0x50, 0x01000010 },
+	{ 0x50, 0x00000010 },
+	{ 0xc0, 0x0e407304 },
+	{ 0xc0, 0x0e447304 },
+	{ 0xc0, 0x0e447306 },
+	{ 0xc0, 0x0e447304 },
+	{ 0xc0, 0x0e407306 },
+};
+
+static struct imx7_pm_data imx7d_pm_data_ddr3 = {
+	.ddrc_num = ARRAY_SIZE(imx7d_ddrc_ddr3_setting),
+	.ddrc_offset = imx7d_ddrc_ddr3_setting,
+	.ddrc_phy_num = ARRAY_SIZE(imx7d_ddrc_phy_ddr3_setting),
+	.ddrc_phy_offset = imx7d_ddrc_phy_ddr3_setting,
+};
+
+paddr_t phys_addr[] = {
+	AIPS1_BASE, AIPS2_BASE, AIPS3_BASE
+};
+
+int pm_imx7_iram_tlb_init(void)
+{
+	uint32_t i;
+	struct tee_mmap_region map;
+
+	/* iram tlb already initialized */
+	if (iram_tlb_phys_addr != (-1UL))
+		return 0;
+
+	iram_tlb_phys_addr = TRUSTZONE_OCRAM_START + 16 * 1024;
+	iram_tlb_virt_addr = phys_to_virt(iram_tlb_phys_addr,
+					  MEM_AREA_TEE_COHERENT);
+
+	/* 16KB */
+	memset(iram_tlb_virt_addr, 0, 16 * 1024);
+
+	for (i = 0; i < ARRAY_SIZE(phys_addr); i++) {
+		map.pa = phys_addr[i];
+		map.va = (vaddr_t)phys_to_virt(phys_addr[i], MEM_AREA_IO_SEC);
+		map.region_size = 0x100000;
+		map.size = AIPS1_SIZE; /* 4M for AIPS1/2/3 */
+		map.type = MEM_AREA_IO_SEC;
+		map.attr = TEE_MATTR_VALID_BLOCK | TEE_MATTR_PRW |
+			   TEE_MATTR_GLOBAL | TEE_MATTR_SECURE |
+			   (TEE_MATTR_CACHE_NONCACHE << TEE_MATTR_CACHE_SHIFT);
+		map_memarea_sections(&map, (uint32_t *)iram_tlb_virt_addr);
+	}
+
+	/* Note IRAM_S_BASE is not 1M aligned, so take care */
+	map.pa = ROUNDDOWN(IRAM_S_BASE, 0x100000);
+	map.va = (vaddr_t)phys_to_virt(map.pa, MEM_AREA_TEE_COHERENT);
+	map.region_size = 0x100000;
+	map.size = 0x100000;
+	map.type = MEM_AREA_TEE_COHERENT;
+	map.attr = TEE_MATTR_VALID_BLOCK | TEE_MATTR_PRWX | TEE_MATTR_GLOBAL |
+		TEE_MATTR_SECURE;
+	map_memarea_sections(&map, (uint32_t *)iram_tlb_virt_addr);
+
+	map.pa = GIC_BASE;
+	map.va = (vaddr_t)phys_to_virt((paddr_t)GIC_BASE, MEM_AREA_IO_SEC);
+	map.region_size = 0x100000;
+	map.size = CORE_MMU_DEVICE_SIZE;
+	map.type = MEM_AREA_TEE_COHERENT;
+	map.attr = TEE_MATTR_VALID_BLOCK | TEE_MATTR_PRW | TEE_MATTR_GLOBAL |
+		TEE_MATTR_SECURE;
+	map_memarea_sections(&map, (uint32_t *)iram_tlb_virt_addr);
+
+
+	/*
+	 * Note: No map DRAM space, DRAM is in auto-selfrefresh,
+	 * If map DRAM in to MMU, mmu may access DRAM which
+	 * hang system.
+	 */
+
+	return 0;
+}
+
+int imx7_suspend_init(void)
+{
+	uint32_t i;
+	uint32_t (*ddrc_offset_array)[2];
+	uint32_t (*ddrc_phy_offset_array)[2];
+	uint32_t suspend_ocram_base = (uint32_t)phys_to_virt(
+			(paddr_t)TRUSTZONE_OCRAM_START +
+			SUSPEND_OCRAM_OFFSET, MEM_AREA_TEE_COHERENT);
```

Comment:
Since this is coherent memory, we can expect phys=virt. Maybe use `phys_to_virt()` only under an `assert()`.

Reply:
I could not get you. Could you please explain more?